// @babylonjs/core/Particles/Node/Enums/nodeParticleContextualSources.js
/* eslint-disable @typescript-eslint/naming-convention */
/**
 * Defines the kind of contextual sources for node particles
 */
export var NodeParticleContextualSources;
(function (NodeParticleContextualSources) {
    /** None */
    NodeParticleContextualSources[NodeParticleContextualSources["None"] = 0] = "None";
    /** Position */
    NodeParticleContextualSources[NodeParticleContextualSources["Position"] = 1] = "Position";
    /** Direction */
    NodeParticleContextualSources[NodeParticleContextualSources["Direction"] = 2] = "Direction";
    /** Age */
    NodeParticleContextualSources[NodeParticleContextualSources["Age"] = 3] = "Age";
    /** Lifetime */
    NodeParticleContextualSources[NodeParticleContextualSources["Lifetime"] = 4] = "Lifetime";
    /** Color */
    NodeParticleContextualSources[NodeParticleContextualSources["Color"] = 5] = "Color";
    /** ScaledDirection */
    NodeParticleContextualSources[NodeParticleContextualSources["ScaledDirection"] = 6] = "ScaledDirection";
    /** Scale */
    NodeParticleContextualSources[NodeParticleContextualSources["Scale"] = 7] = "Scale";
    /** AgeGradient */
    NodeParticleContextualSources[NodeParticleContextualSources["AgeGradient"] = 8] = "AgeGradient";
    /** Angle */
    NodeParticleContextualSources[NodeParticleContextualSources["Angle"] = 9] = "Angle";
    /** SpriteCellIndex */
    NodeParticleContextualSources[NodeParticleContextualSources["SpriteCellIndex"] = 16] = "SpriteCellIndex";
    /** SpriteCellStart */
    NodeParticleContextualSources[NodeParticleContextualSources["SpriteCellStart"] = 17] = "SpriteCellStart";
    /** SpriteCellEnd */
    NodeParticleContextualSources[NodeParticleContextualSources["SpriteCellEnd"] = 18] = "SpriteCellEnd";
    /** Initial Color */
    NodeParticleContextualSources[NodeParticleContextualSources["InitialColor"] = 19] = "InitialColor";
    /** Color Dead*/
    NodeParticleContextualSources[NodeParticleContextualSources["ColorDead"] = 20] = "ColorDead";
    /** Initial Direction */
    NodeParticleContextualSources[NodeParticleContextualSources["InitialDirection"] = 21] = "InitialDirection";
    /** Color Step */
    NodeParticleContextualSources[NodeParticleContextualSources["ColorStep"] = 22] = "ColorStep";
    /** Scaled Color Step */
    NodeParticleContextualSources[NodeParticleContextualSources["ScaledColorStep"] = 23] = "ScaledColorStep";
    /** Local Position Updated */
    NodeParticleContextualSources[NodeParticleContextualSources["LocalPositionUpdated"] = 24] = "LocalPositionUpdated";
    /** Size */
    NodeParticleContextualSources[NodeParticleContextualSources["Size"] = 25] = "Size";
    /** Direction Scale */
    NodeParticleContextualSources[NodeParticleContextualSources["DirectionScale"] = 32] = "DirectionScale";
})(NodeParticleContextualSources || (NodeParticleContextualSources = {}));
